data-prepper-plugins/otel-trace-raw-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltrace/OTelTraceRawPrepper.java

```diff
@@ -197,13 +209,22 @@ private void processDescendantRawSpan(final RawSpan rawSpan) {
                     Iterator<Map.Entry<String, RawSpanSet>> entryIterator = traceIdRawSpanSetMap.entrySet().iterator();
                     while (entryIterator.hasNext()) {
                         Map.Entry<String, RawSpanSet> entry = entryIterator.next();
+                        String traceId = entry.getKey();
+                        String traceGroup = traceIdTraceGroupCache.get(traceId);
                         RawSpanSet rawSpanSet = entry.getValue();
                         long traceTime = rawSpanSet.getTimeSeen();
                         if (now - traceTime >= traceFlushInterval) {
                             Set<RawSpan> rawSpans = rawSpanSet.getRawSpans();
-                            for (RawSpan rawSpan : rawSpans) {
-                                recordsToFlush.add(rawSpan);
-                                log.warn("Missing root span for SpanId: {}", rawSpan.getSpanId());
+                            if (traceGroup != null) {
+                                rawSpans.forEach(rawSpan -> {
+                                    rawSpan.setTraceGroup(traceGroup);
+                                    recordsToFlush.add(rawSpan);
+                                });
+                            } else {
+                                rawSpans.forEach(rawSpan -> {
+                                    recordsToFlush.add(rawSpan);
+                                    log.warn("Missing root span for SpanId: {}", rawSpan.getSpanId());
```

Comment:
Might be more accurate to say "missing trace group for SpanId" (if there's ever a case where we have the root span yet trace group is still somehow null).